src/aspire/source/__init__.py

```diff
@@ -370,6 +375,120 @@ def whiten(self, noise_filter):
         # Invalidate images
         self._im = None
 
+    def phase_flip(self):
+        """
+        Perform phase flip to images in the source object using CTF information
+        """
+        logger.info('Perform phase flip on source object')
+        logger.info('Adding Phase Flip Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(FlipXform(self.filters))
+        # Invalidate images
+        self._im = None
+
+    def invert_contrast(self, batch_size=512):
+        """
+        invert the global contrast of images
+
+        Check if all images in a stack should be globally phase flipped so that
+        the molecule corresponds to brighter pixels and the background corresponds
+        to darker pixels. This is done by comparing the mean in a small circle
+        around the origin (supposed to correspond to the molecule) with the mean
+        of the noise, and making sure that the mean of the molecule is larger.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending a `Multiple` filter to the generation pipeline.
+
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info('Apply contrast inversion on source object')
+        L = self.L
+        grid = grid_2d(L, shifted=True)
+        # Get mask indices of signal and noise samples assuming molecule
+        signal_mask = (grid['r'] < 0.5)
+        noise_mask = (grid['r'] > 0.8)
+
+        # Calculate mean values in batch_size
+        signal_mean = 0.0
+        noise_mean = 0.0
+
+        for i in range(0, self.n, batch_size):
+            images = self.images(i, batch_size).asnumpy()
+            signal = images * np.expand_dims(signal_mask, 2)
+            noise = images * np.expand_dims(noise_mask, 2)
+            signal_mean += np.sum(signal)
+            noise_mean += np.sum(noise)
+        signal_denominator = self.n * np.sum(signal_mask)
+        noise_denominator = self.n * np.sum(noise_mask)
+        signal_mean /= signal_denominator
+        noise_mean /= noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to invert contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to invert contrast')
+            scale_factor = 1.0 * np.ones(self.n)
```

Comment:
A huge array of a single constant value usually indicates an underlying design issue - in this case that `Multiply` Xform insists on an `ndarray`, and cannot deal with a scaler multiplier. Can you open a new non-priority issue for this (and reference this line).

Reply:
Issue #233 was created for this and Shift Xform too.